hudi-client/hudi-spark-client/src/main/java/org/apache/hudi/table/action/commit/ConsistentBucketBulkInsertDataInternalWriterHelper.java

```diff
@@ -0,0 +1,120 @@
+import org.apache.hudi.table.action.cluster.util.ConsistentHashingUpdateStrategyUtils;
+
+import org.apache.spark.sql.catalyst.InternalRow;
+import org.apache.spark.sql.types.StructType;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.util.Collections;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+/**
+ * Helper class for native row writer for bulk_insert with consistent hashing bucket index.
+ */
+public class ConsistentBucketBulkInsertDataInternalWriterHelper extends BucketBulkInsertDataInternalWriterHelper {
+
+  private static final Logger LOG = LoggerFactory.getLogger(ConsistentBucketBulkInsertDataInternalWriterHelper.class);
+
+  public ConsistentBucketBulkInsertDataInternalWriterHelper(HoodieTable hoodieTable, HoodieWriteConfig writeConfig,
+                                                            String instantTime, int taskPartitionId, long taskId, long taskEpochId, StructType structType,
+                                                            boolean populateMetaFields, boolean arePartitionRecordsSorted) {
+    this(hoodieTable, writeConfig, instantTime, taskPartitionId, taskId, taskEpochId, structType, populateMetaFields, arePartitionRecordsSorted, false);
+  }
+
+  public ConsistentBucketBulkInsertDataInternalWriterHelper(HoodieTable hoodieTable, HoodieWriteConfig writeConfig,
+                                                            String instantTime, int taskPartitionId, long taskId, long taskEpochId, StructType structType,
+                                                            boolean populateMetaFields, boolean arePartitionRecordsSorted, boolean shouldPreserveHoodieMetadata) {
+    super(hoodieTable, writeConfig, instantTime, taskPartitionId, taskId, taskEpochId, structType, populateMetaFields, arePartitionRecordsSorted, shouldPreserveHoodieMetadata);
+  }
+
+  public void write(InternalRow row) throws IOException {
+    try {
+      if (handle == null) {
+        // We can ensure that one writer has only one handle
+        String partitionPath = String.valueOf(extractPartitionPath(row));
+        String recordKey = String.valueOf(extractRecordKey(row));
+        handle = getBucketRowCreateHandle(partitionPath, recordKey);
+      }
+      handle.write(row);
+    } catch (Throwable t) {
+      LOG.error("Global error thrown while trying to write records in HoodieRowCreateHandle ", t);
+      throw new IOException(t);
+    }
+  }
+
+  private HoodieRowCreateHandle getBucketRowCreateHandle(String partitionPath, String recordKey) {
+    ConsistentBucketIdentifier identifier = getBucketIdentifier(partitionPath);
+    final ConsistentHashingNode node = identifier.getBucket(recordKey, indexKeyFields);
+    String fileId = FSUtils.createNewFileId(node.getFileIdPrefix(), 0);
+
+    ValidationUtils.checkArgument(node.getTag() != ConsistentHashingNode.NodeTag.NORMAL
+            || hoodieTable.getFileSystemView()
```

Comment:
Is this check too heavy to do when starting write to a new file group? 

I think it might be simpler if we add a file existing check in the init method of class `HoodieRowCreateHandle` and fail the writer if file existing.

e.g.

```java
  public HoodieRowCreateHandle(HoodieTable table,
                               HoodieWriteConfig writeConfig,
                               String partitionPath,
                               String fileId,
                               String instantTime,
                               int taskPartitionId,
                               long taskId,
                               long taskEpochId,
                               StructType structType,
                               boolean shouldPreserveHoodieMetadata) {
    ...
    FileSystem fs = table.getMetaClient().getFs();

    String writeToken = getWriteToken(taskPartitionId, taskId, taskEpochId);
    String fileName = FSUtils.makeBaseFileName(instantTime, writeToken, this.fileId, table.getBaseFileExtension());
    this.path = makeNewPath(fs, partitionPath, fileName, writeConfig);
    if (fs.exists(this.path)) {
      throw new HoodieException("File already exist. Can't bulk insert. Can use insert|upsert"); 
    }
    ...
```

Reply:
Can one writer helper writes multiple file groups?

Reply:
Yes, one writer helper can write multiple file groups, but one object of `HoodieRowCreateHandle` can only write one parquet file, which is `this.path`. This write helper will switch between the different handlers.

So in the method `BulkInsertDataInternalWriterHelper.write(InternalRow row)`, `BucketBulkInsertDataInternalWriterHelper.write(InternalRow row)` and `ConsistentBucketBulkInsertDataInternalWriterHelper.write(InternalRow row)`, they will create a new object of `HoodieRowCreateHandle` if the incoming row need to be written into a new parquet file.

We only have this existing file group check in `ConsistentBucketBulkInsertDataInternalWriterHelper.getBucketRowCreateHandle`, but actually, other bulk insert writer `BulkInsertDataInternalWriterHelper` and `BucketBulkInsertDataInternalWriterHelper` need this check as well. So I think it will be simpler if we add a check in the init method of `HoodieRowCreateHandle`.

Reply:
And another thing, I think it is a bug in `ConsistentBucketBulkInsertDataInternalWriterHelper` that we only create a new object of `HoodieRowCreateHandle` in the method `ConsistentBucketBulkInsertDataInternalWriterHelper.getBucketRowCreateHandle`, but didn't put the object in the map `handles`.

correct me if i am wrong.

Reply:
Oh sorry, my bad, just check the code, this writer help class is used in the method `HoodieDatasetBulkInsertHelper.bulkInsert` in a `mapPartitions` method, so one writer helper only write one file.

Reply:
If it looks reasonable for improvement, you can try to raise a PR then, maybe @stream2000 can help for a reivew~

Reply:
Yeah sure, I'm adding a few test in our internal code, will raise a PR for with the test cases and fix.

Reply:
@TengHuo I think using file system view which can help filter out replaced/uncommitted file slices is a more common way when we need information about file slices. And does the write token is the same between every write for the same file id?

Reply:
Oh got it, Indeed, `writeToken` is different every time when a task or stage failed and retry. So, just checking `fs.exists(this.path)` is not good enough, need to check file group with `instantTime`. In that case, using `hoodieTable.getFileSystemView()` is better then list partition.